draft-ietf-protected-audience-key-value-services.md

```diff
@@ -0,0 +1,717 @@
+  AEAD-Protected Response (..),
+}
+~~~~~
+
+
+The response uses the a similar encapsulated response format to that used by [OHTTP] (see
+{{Section 4.4 of OHTTP}}), but with the custom `message/ad-auction-trusted-signals-response`
+media type instead of `message/bhttp response`:
+
+1. Export a secret (`secret`) from `context`, using the string
+   "message/ad-auction-trusted-signals-response" as the `exporter_context` parameter to
+   `context.Export`; see [Section 5.3](https://www.rfc-editor.org/rfc/rfc9180.html#name-secret-export)
+   of [HPKE]. The length of this secret is `max(Nn, Nk)`, where `Nn` and `Nk` are
+   the length of the AEAD key and nonce that are associated with `context`.
+2. Generate a random value of length `max(Nn, Nk)` bytes, called `response_nonce`.
+3. Extract a pseudorandom key (`prk`) using the `Extract` function provided by
+   the KDF algorithm associated with context. The `ikm` input to this function
+   is `secret`; the `salt` input is the concatenation of `enc` (from
+  `enc_request`) and `response_nonce`.
+1. Use the `Expand` function provided by the same KDF to create an AEAD key,
+   `key`, of length `Nk` -- the length of the keys used by the AEAD associated
+   with `context`. Generating `aead_key` uses a label of "key".
+2. Use the same `Expand` function to create a nonce, `nonce`, of length `Nn`
+   -- the length of the nonce used by the AEAD. Generating `aead_nonce` uses a
+  label of "nonce".
+1. Encrypt `response`, passing the AEAD function `Seal` the values of `aead_key`,
+   `aead_nonce`, an empty `aad`, and a `pt` input of `response`. This yields `ct`.
+2. Concatenate `response_nonce` and `ct`, yielding an Encapsulated Response,
+   `enc_response`. Note that `response_nonce` is of fixed length, so there is no
+  ambiguity in parsing either `response_nonce` or `ct`.
+
+In pseudocode, this procedure is as follows:
+
+~~~~~
+secret = context.Export("message/ad-auction-trusted-signals-response", max(Nn, Nk))
+response_nonce = random(max(Nn, Nk))
+salt = concat(enc, response_nonce)
+prk = Extract(salt, secret)
+aead_key = Expand(prk, "key", Nk)
+aead_nonce = Expand(prk, "nonce", Nn)
+ct = Seal(aead_key, aead_nonce, "", response)
+enc_response = concat(response_nonce, ct)
+~~~~~
+
+Clients decrypt an Encapsulated Response by reversing this process. That is,
+Clients first parse `enc_response` into `response_nonce` and `ct`. Then, they
+follow the same process to derive values for `aead_key` and `aead_nonce`, using
+their sending HPKE context, `sctxt`, as the HPKE context, `context`.
+
+The Client uses these values to decrypt `ct` using the AEAD function `Open`.
+Decrypting might produce an error, as follows:
+
+~~~~~
+response, error = Open(aead_key, aead_nonce, "", ct)
+~~~~~
+
+### Framing and Padding {#response-framing}
+
+The plaintext response message uses the framing described in {{framing}}.
+
+Padding is applied with sizes as multiples of 2^n KBs ranging from 0 to 2MB. So the valid response
```

Comment:
What does the server do if the response does not fit in 2MB?

Reply:
@peiwenhu

Reply:
It returns an error response code. in the future the response code can be inside the ciphertext once we have error handling defined.